Fix "Only X left" functionality

aleron75 opened the pull request:
This PR is intended to fix issue #1165 

- Add `InventoryFrontendUi` module to implement the fix
- Pluginize methods used by `\Magento\CatalogInventory\Block\Stockqty\AbstractStockqty` instances
- Update `InventoryConfiguration` to access the stock threshold qty 

This PR is also intended to fix issue #1349

aleron75 committed:
Fix "Only X left" functionality

- Add InventoryFrontendUi module to implement the fix for issue 1165
- Pluginize methods used by \Magento\CatalogInventory\Block\Stockqty\AbstractStockqty instances
- Update InventoryConfiguration to access the stock threshold qty

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=1348) <br/>All committers have signed the CLA.

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+ * Created by PhpStorm.
+ * User: alessandro
+ * Date: 09/06/18
+ * Time: 15.35
```
please fix the PHP Doc Block

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+     */
+    private $getProductSalableQty;
+
+    public function __construct(
```
please provide doc block for constructor

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+        $this->getProductSalableQty = $getProductSalableQty;
+    }
+
+    /**
```
please specify parameter for this method in doc block

maghamed commented on app/code/Magento/InventoryFrontendUi/composer.json:
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "magento/module-inventory-frontend-ui",
```
it's better to rename module to InventorySalesFrontendUi

maghamed commented on app/code/Magento/InventoryFrontendUi/composer.json:
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "magento/module-inventory-frontend-ui",
+    "description": "N/A",
+    "require": {
```
please add the reference to this module in "require" section of the root composer.json

maghamed commented on app/code/Magento/InventoryFrontendUi/etc/module.xml:
```diff
@@ -0,0 +1,10 @@
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Module/etc/module.xsd">
+    <module name="Magento_InventoryFrontendUi" setup_version="1.0.0" />
```
no need to use `setup_version` 

that's true for other Inventory modules as well

maghamed commented on app/code/Magento/InventoryFrontendUi/etc/di.xml:
```diff
@@ -0,0 +1,15 @@
+    <type name="Magento\CatalogInventory\Block\Stockqty\AbstractStockqty">
+        <plugin name="magentoInventoryFrontendUiAbstractStockqty"
+                type="Magento\InventoryFrontendUi\Plugin\Block\Stockqty\AbstractStockqtyPlugin"/>
+        <plugin name="magentoInventoryFrontendUiAbstractStockqty"
```
looks like it's copy-paste of previous string

maghamed commented on app/code/Magento/InventoryFrontendUi/composer.json:
```diff
@@ -0,0 +1,25 @@
+        "magento/framework": "*",
+        "magento/module-inventory-api": "*",
+        "magento/module-inventory-sales": "*",
+        "magento/module-inventory-configuration": "*",
```
dependency should be fixed on `module-inventory-configuration-api`

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+        $minStockQty = $stockItemConfig->getMinQty();
+        return $this->getProductSalableQty->execute($sku, $stockId) - $minStockQty;
+    }
+}
```
static tests will break here

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+ * Date: 09/06/18
+ * Time: 15.35
+ */
+namespace Magento\InventoryFrontendUi\Plugin\Block\Stockqty;
```
there is no declare strict type here

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    protected function getStockQtyLeft(string $sku, int $stockId): float
```
protected methods are forbidden in newly created code, just private is allowed to use

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+
+    /**
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
```
provide correct doc type

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+            return $this->getProductSalableQty->execute($sku, $stockId);
+        }
+        $minStockQty = $stockItemConfig->getMinQty();
+        return $this->getProductSalableQty->execute($sku, $stockId) - $minStockQty;
```
let's add `$stockItemConfig->getMinQty()` into the `getProductSalableQty` service

so that we can substitute getStockQtyLeft
with getProductSalableQuantity

aleron75 replied:
Since incorporating the min qty subtraction into the `getProductSalableQty` service requires refactoring of different pieces of code, I opened the issue #1349, assigned it to myself, not to block this PR.

maghamed commented on app/code/Magento/InventoryFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+        if (null === $stockItemConfig) {
+            return false;
+        }
+        return $this->getProductSalableQty->execute($sku, $stockId) > 0
```
let's use clean check for Backorder
`getBackorders` instead of `$this->getProductSalableQty->execute($sku, $stockId) > 0`
because even if we have Salable Qty = 5 , and getStockThresholdQty = 10
that's not correct to say that only 5 left, because having backorder enabled we can sell infinite number

aleron75 replied:
I'm not sure I understood your proposal, I changed the code like this:
```
return $stockItemConfig->getBackorders() === StockItemConfigurationInterface::BACKORDERS_NO
            && $this->getProductSalableQty->execute($sku, $stockId) > 0
            && $this->getStockQtyLeft($sku, $stockId) <= $stockItemConfig->getStockThresholdQty();
```
If backorders are enabled, the first condition will fail and the message won't be dispayed.

If backorders are disabled, the first condition will return `true` and the other two conditions check that the product is salable and that the salable qty is lett or equal than the threshold.

Is that correct?

aleron75 committed:
Apply code changes according to PR review

- Rename module to `InventorySalesFrontendUi`
- Fix several coding style issues
- Take backorders into account to dispay or not the "Only X left" message

aleron75 committed:
Update dependencies and composer.lock

aleron75 committed:
Remove reduntant dependencies

aleron75 committed:
Apply refactoring to addres issue #1349

- Update GetProductSalableQty service to subtract min qty
- Update code that use the GetProductSalableQty service and min qty

aleron75 committed:
Rename variable

maghamed commented on app/code/Magento/InventorySalesApi/Api/GetProductSalableQtyInterface.php:
```diff
@@ -20,6 +22,7 @@
      * @param string $sku
      * @param int $stockId
      * @return float
+     * @throws \Magento\Framework\Exception\LocalizedException
```
also need to add InputExcepton here 
as implementation may throw 
```
            throw new InputException(
                __('Can\'t check requested quantity for products without Source Items support.')
            );
```

maghamed commented on app/code/Magento/InventorySalesFrontendUi/Plugin/Block/Stockqty/AbstractStockqtyPlugin.php:
```diff
@@ -0,0 +1,86 @@
+            return false;
+        }
+        return $stockItemConfig->getBackorders() === StockItemConfigurationInterface::BACKORDERS_NO
+            && $this->getProductSalableQty->execute($sku, $stockId) > 0
```
we don't need this check 
` && $this->getProductSalableQty->execute($sku, $stockId) > 0`

because in this case product would not be salable at all

aleron75 committed:
Add message vibibility check based on product type

aleron75 committed:
Fix code comment and remove useless check for qty > 0

aleron75 committed:
Merge branch '2.3-develop' into msi-1165-fix-only-x-left-functionality

smoskaluk commented:
@aleron75 @maghamed I could not verify this PR because message 'Only x left' appear on product page only after flushing the cache. Please re-check
